tests/unit_tests/feature/testgetfeaturelistscheduler.cpp

```diff
@@ -0,0 +1,160 @@
+  for (int i = 0; i < 2; ++i) {
+    const TaskGetFeatureList* task =
+        dynamic_cast<const TaskGetFeatureList*>(updatedListOfTasks.takeLast());
+    // If it is not a nullptr, the cast was succesfull.
+    QVERIFY(task);
+  }
+
+  // Remove the TaskScheduler singleton so that state doesn't leak to other
+  // tests.
+  TaskScheduler::reset();
+}
+
+void TestGetFeatureListScheduler::testTaskIsScheduledOnTokenChange() {
+  SettingsHolder settingsHolder;
+
+  // Stop the TaskScheduler so we can inpect the tasks scheduled.
+  TaskScheduler::stop();
+
+  auto initialListOfTasks = TaskScheduler::tasks();
+
+  GetFeatureListScheduler worker;
+  worker.start(1 * 1000 * 60 * 60);  // 1hr
+
+  settingsHolder.setToken("aToken");
+  settingsHolder.setToken("anotherToken");
+  settingsHolder.removeToken();
+
+  auto updatedListOfTasks = TaskScheduler::tasks();
+
+  // We are expecting four tasks, the task scheduled on start + 3.
+  QCOMPARE(initialListOfTasks.count() + 4, updatedListOfTasks.count());
+
+  // Check that the tasks are TaskGetFeatureList, just in case.
+  for (int i = 0; i < 4; ++i) {
+    const TaskGetFeatureList* task =
+        dynamic_cast<const TaskGetFeatureList*>(updatedListOfTasks.takeLast());
+    // If it is not a nullptr, the cast was succesfull.
+    QVERIFY(task);
+  }
+
+  // Remove the TaskScheduler singleton so that state doesn't leak to other
+  // tests.
+  TaskScheduler::reset();
+}
+
+void TestGetFeatureListScheduler::testTimerIsStoppedOnDestruction() {
+  SettingsHolder settingsHolder;
+
+  // Stop the TaskScheduler so we can inpect the tasks scheduled.
+  TaskScheduler::stop();
+
+  auto initialListOfTasks = TaskScheduler::tasks();
+
+  {
+    GetFeatureListScheduler worker;
+    worker.start(200);  // 200ms
+
+    // Sleep for 250 milliseconds,
+    // give the worker time to schedule the first periodic task.
+    QTest::qWait(250);
+
+    auto updatedListOfTasks = TaskScheduler::tasks();
+
+    // We are expecting two tasks, the task scheduled on start + 1.
+    QCOMPARE(initialListOfTasks.count() + 2, updatedListOfTasks.count());
+
+    // Check that the tasks are TaskGetFeatureList, just in case.
+    // Check that the tasks are TaskGetFeatureList, just in case.
+    for (int i = 0; i < 2; ++i) {
+      const TaskGetFeatureList* task = dynamic_cast<const TaskGetFeatureList*>(
+          updatedListOfTasks.takeLast());
+      qDebug() << updatedListOfTasks.count();
+      // If it is not a nullptr, the cast was succesfull.
+      QVERIFY(task);
+    }
+
+    // Update the initial list of tasks before destruction of the worker.
+    initialListOfTasks = TaskScheduler::tasks();
+  }
+
+  // Sleep more 250, if the timer hasn't stopped this will schedule more
+  // tasks. But it shouldn't.
+  QTest::qWait(250);
+
+  // Also change the token, this should also not schedule any tasks unless
+  // the signal connection was done incorrectly.
+  settingsHolder.setToken("aToken");
```

Comment:
Should this token be removed in test cleanup?

Reply:
The settings holder is destructed at the end of this test case and that takes care of it already.

Reply:
But doesn't setToken write the token to the backing settings file? How does the SettingHolder destruction help with that?

Reply:
Correct, however the SettingsHolder destructor will also clean that for tests: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/settingsholder.cpp#L64-L68. This is a bit hacky, but I did it so I did not have to touch most tests that already expect this to be the default behaviour.